util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+        NULL
+      );
+      if (handle == INVALID_HANDLE_VALUE) {
+        return WindowsError(filename_, GetLastError());
+      }
+    }
+
+    pread_mutex.Lock();
+
+    // Seek offset
+    Status s = Status::OK();
+    LARGE_INTEGER in, out;
+    in.QuadPart = offset;
+    if (!SetFilePointerEx(handle, in, &out, FILE_BEGIN)) {
+      s = WindowsError(filename_, GetLastError());
+    }
+
+    // read from file
+    DWORD r = 0;
+    if (s.ok() && !::ReadFile(handle, scratch, n, &r, NULL)) {
+      s = WindowsError(filename_, GetLastError());
+    }
+
+    pread_mutex.Unlock();
+
+    *result = Slice(scratch, r);
+    if (temporary_handle_) {
+      // Close the temporary file descriptor opened earlier.
+      CloseHandle(handle);
+    }
+    return s;
+  }
+};
+
+
+class WindowsMmapReadableFile: public RandomAccessFile
+{
+ private:
+  std::string filename_;
+  void* mmapped_region_;
+  size_t length_;
+  Limiter* limiter_;
+  HANDLE filemap_;
+
+ public:
+  // base[0,length-1] contains the mmapped contents of the file.
+  WindowsMmapReadableFile(const std::string& fname, void* base, size_t length,
+                          HANDLE filemap, Limiter* limiter)
+      : filename_(fname), mmapped_region_(base), length_(length),
+        limiter_(limiter), filemap_(filemap) {
+  }
+
+  virtual ~WindowsMmapReadableFile() {
+    ::UnmapViewOfFile(mmapped_region_);
+    CloseHandle(filemap_);
+    limiter_->Release();
+  }
+
+  virtual Status Read(uint64_t offset, size_t n, Slice* result,
+                      char* scratch) const {
+    Status s;
+    if (offset + n > length_) {
+      *result = Slice();
+      s = WindowsError(filename_, ERROR_INVALID_PARAMETER);
+    } else {
+      *result = Slice(reinterpret_cast<char*>(mmapped_region_) + offset, n);
+    }
+    return s;
+  }
+};
+
+
+class WindowsWritableFile : public WritableFile {
+ private:
+  std::string filename_;
+  HANDLE handle_;
+
+ public:
+  WindowsWritableFile(const std::string& fname, HANDLE handle)
+      : filename_(fname), handle_(handle) { }
+
+  virtual ~WindowsWritableFile() {
+    if (handle_ != INVALID_HANDLE_VALUE) {
+      // Ignoring any potential errors
+      CloseHandle(handle_);
+    }
+  }
+
+  virtual Status Append(const Slice& data) {
+    DWORD wrote = 0;
+    if (!::WriteFile(handle_, data.data(), data.size(), &wrote, NULL)) {
+      return WindowsError(filename_, GetLastError());
+    }
+    if (wrote != data.size()) {
+      return Status::IOError(filename_, "Could not write entire buffer to disk");
+    }
+    return Status::OK();
+  }
+
+  virtual Status Close() {
+    Status result;
+    if (!::CloseHandle(handle_)) {
+      result = WindowsError(filename_, GetLastError());
+    }
+    handle_ = INVALID_HANDLE_VALUE;
+    return result;
+  }
+
+  virtual Status Flush() {
+    if (!::FlushFileBuffers(handle_)) {
+      return WindowsError(filename_, GetLastError());
+    }
+    return Status::OK();
+  }
+
+  Status SyncDirIfManifest() {
```

Comment:
Since SyncDirIfManifest() is empty, just delete it and move the comment into Sync.